text/0000-offset_of.md

```diff
@@ -0,0 +1,482 @@
+
+However, if a `#[repr(...)]` is not used, the compiler is free to place the
+fields of `Vertex` in whatever order it prefers (even if they aren't the same as
+the order the fields are written in the struct declaration), so there's no way
+to know in advance what the positions of the fields will be.
+
+Thankfully, `offset_of!` is still usable here:
+
+```rs
+// No `#[repr()]` needed!
+struct Vertex {
+    tex: [u16; 2],
+    pos: [f32; 3],
+}
+// This time let's define some constants containing the offset value,
+// which can be more readable if you need to use them several times.
+const OFFSET_VERTEX_TEX: usize = core::mem::offset_of!(Vertex, tex);
+const OFFSET_VERTEX_POS: usize = core::mem::offset_of!(Vertex, pos);
+```
+
+As you can see, the usage is the same as before, but because we didn't specify
+`#[repr(C)]`, compiler may have changed the order or position, so the values may
+be different -- it's completely possible that `pos` is located at offset 0, for
+example! Thankfully, by using `core::mem::offset_of!`, this code is correct
+either way, and will continute to be correct, even if the layout algorithm
+changes in the future.
+
+## `offset_of!` On Other Types
+
+If your type doesn't have named fields, `offset_of!` can still be used. For
+tuples and tuple structs, the "name" of the field is the numeral value you use
+to access it. For example:
+
+```rust
+// Works with a tuple struct
+struct KeyVal(&'static str, Vec<u8>);
+const OFFSET_KV_KEY: usize = core::mem::offset_of!(KeyVal, 0);
+const OFFSET_KV_VAL: usize = core::mem::offset_of!(KeyVal, 1);
+// Or with an anonymous tuple.
+const OFFSET_ANON_KEY: usize = core::mem::offset_of!((&'static str, Vec<u8>), 0);
+const OFFSET_ANON_VAL: usize = core::mem::offset_of!((&'static str, Vec<u8>), 1);
+```
+
+Finally, `offset_of!` can be used to compute the offset of fields in unions too.
+While this may be surprising, the compiler is allowed to put padding in front of
+fields in unions which are not `#[repr(C)]`, which would lead to a non-zero
+field offset.
+
+```rs
+use core::mem::offset_of;
+union Buffer {
+    metadata: [u64; 3],
+    datadata: [u8; 1024 * 1024 * 32],
+}
+const METADATA_OFFSET: usize = offset_of!(Buffer, metadata);
+```
+
+## Limitations
+
+There are a few limitations worth mentioning. Some of these may be relaxed in
+the future, however.
+
+1. Perhaps unsurprisingly, it obeys privacy, so both the type and field you call
+   `offset_of!` on must be visible to the code calling `offset_of!`.
+
+2. The type holding the field must be `Sized`, so trying to compute where the
+   slice begins in something like `offset_of!((i32, [u32]), 1)` isn't supported.
+
+3. Compared to `offsetof` in C and C++, you can't access nested fields/arrays.
+   That is, instead of `offset_of!(Foo, quank.zoop.2.quank[4])`, you'll have to
+   compute the offsets of each step manually, and sum them.
+
+4. Finally, types other than tuples, structs, and unions are currently
```

Comment:
What about arrays? It's pretty easy to compute, but it seems like it would make sense to allow something like `offset_of!([T; N], 2)` as a shorthand for 2 times the size of `T` rounded up to the alignment of `T`.

Reply:
Size is always a multiple of alignment, so this is just `2 * size_of::<T>()`. I suppose we could allow it for consistency, but I don't really see a good motivation tbh

Reply:
This isn't a field so I'm really not a fan of allowing this. I kind of hint at it in the nested fields part of the future work, though. I think without nested field access, it's not worth the inconsistency that it introduces.